sos/report/plugins/mellanox_firmware.py

```diff
@@ -0,0 +1,141 @@
+            return
+
+        self.collect_cmd_output('mst cable add')
+        self.collect_cmd_output("mst status -v", timeout=10)
+        self.collect_cmd_output("mlxcables", timeout=10)
+        co = os.listdir("/dev/mst")
+        mlxcables = []
+        for device in co:
+            if 'cable' in device:
+                mlxcables.append(device)
+        for mlxcable in mlxcables:
+            self.collect_cmd_output(f"mlxcables -d {mlxcable} --DDM",
+                                    timeout=10)
+            self.collect_cmd_output(f"mlxcables -d {mlxcable} --dump",
+                                    timeout=10)
+        self.collect_cmd_output("mst stop", changes=True)
+
+    def setup(self):
+        # Get all devices which have the vendor Mellanox Technologies
+        devices = []
+        device_list = self.collect_cmd_output('lspci -D -d 15b3::0200')
+        """
+        Will return a string of the following format:
+        0000:08:00.0 Ethernet controller: Mellanox Technologies MT2892
+        Family
+        """
+        if device_list['status'] != 0:
+            # bail out if there no Mellanox PCI devices
+            return
+
+        for line in device_list["output"].splitlines():
+            """
+            Should return 0000:08:00.0
+            from the following string
+            0000:08:00.0 Ethernet controller: Mellanox Technologies MT2892
+            Family
+            """
+            devices.append(line[0:8]+'00.0')
+
+        devices = set(devices)
+
+        """
+        # Mft package is present if OFED is installed
+        # mstflint package is part of the distro and can be installed.
+        """
+        commands = []
+
+        # mft package is installed if flint command is available
+        co = self.exec_cmd('flint --version')
+        if co['status'] != 0:
+            """
+            mstflint package commands
+            the commands do not support position independent arguments
+            """
+            commands = [
+                ["mstconfig -d ", " -e q"],
+                ["mstflint -d ", " dc"],
+                ["mstflint -d ", " q"],
+                ["mstreg -d ", " --reg_name ROCE_ACCL --get"],
+                ["mstlink -d ", ""],
+            ]
+        else:
+            """
+            mft package commands
+            the commands do not support position independent arguments
+            """
+            commands = [
+                ["mlxdump -d ", " pcie_uc --all"],
+                ["mstconfig -d ", " -e q"],
+                ["flint -d ", " dc"],
+                ["flint -d ", " q"],
+                ["mlxreg -d ", " --reg_name ROCE_ACCL --get"],
+                ["mlxlink -d ", ""],
+                ["fwtrace -d ", " -i all --tracer_mode FIFO"],
+            ]
+        for device in devices:
+            for command in commands:
+                self.add_cmd_output(f"{command[0]} {device} "
+                                    f"{command[1]}", timeout=30)
```

Comment:
I'm trying to think of a cleaner way to structure this, but not entirely sure doing so buys us much. Let me think on this a bit, or if @pmoravec has some ideas, but if I/you/we can't come up with an easier-to-grok in a few days then I'm fine with this.

Reply:
Ack. To be fair this looks a bit better than before...
Because the commands accept the device parameter only in that order, it makes things harder than it should be.

Reply:
Yeah I spent a while during my first review here, but havent figured out a better code.